src/mem3/src/mem3_reshard_job.erl

```diff
@@ -170,20 +170,33 @@ run(#job{split_state = CurrState} = Job) ->
         topoff3 -> topoff;
         _ -> CurrState
     end,
-    Job1 = ?MODULE:StateFun(Job),
-    Job2 = wait_for_workers(Job1),
-    Job3 = switch_to_next_state(Job2),
-    ok = checkpoint(Job3),
-    run(Job3).
+    NewJob = try
+        Job1 = ?MODULE:StateFun(Job),
+        Job2 = wait_for_workers(Job1),
+        Job3 = switch_to_next_state(Job2),
+        ok = checkpoint(Job3),
+        Job3
+    catch
+        throw:{retry, RetryJob} ->
+            RetryJob
+    end,
+    run(NewJob).
 
 
 set_start_state(#job{split_state = State} = Job) ->
-    case maps:get(State, ?STATE_RESTART, undefined) of
-        undefined ->
-            Fmt = "~p recover : unknown state ~s",
-            couch_log:error(Fmt, [?MODULE, jobfmt(Job)]),
+    case {State, maps:get(State, ?STATE_RESTART, undefined)} of
+        {_, undefined} ->
+            Fmt1 = "~p recover : unknown state ~s",
+            couch_log:error(Fmt1, [?MODULE, jobfmt(Job)]),
             erlang:error({invalid_split_job_recover_state, Job});
-        StartState ->
+        {initial_copy, initial_copy} ->
+            % We recover from initial_copy to initial_copy we need
```

Comment:
This comment is a bit difficult to parse. Maybe better as:
"_Since_  We recover from initial_copy to initial_copy we need..." or perhaps
" We recover from initial_copy to initial_copy _so_ we need..."?

Reply:
reworded and added two tests